src/Microsoft.OData.Core/UriParser/Resolver/CaseInsensitiveSchemaElementsCache.cs

```diff
@@ -0,0 +1,136 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="CaseInsensitiveSchemaElementsCache.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.OData.Edm
+{
+    /// <summary>
+    /// Cache used to store schema elements using case-normalized names
+    /// to speed up case-insensitive model lookups. The cache is populated
+    /// up front so that if an item is not found in the cache, we can assume
+    /// it doesn't exist in the model. For this reason, it's important that
+    /// the model be immutable.
+    /// </summary>
+    internal sealed class CaseInsensitiveSchemaElementsCache
+    {
+        private static readonly List<IEdmSchemaElement> emptyList = new List<IEdmSchemaElement>();
+        // This cache is meant to be populate up front and remain read-only
+        // after that. Therefore, it doesn't need
+        // to be a concurrent dictionary.
+        private readonly Dictionary<string, List<IEdmSchemaElement>> cache;
+
+        /// <summary>
+        /// Builds a case-insensitive cache of schema elements from
+        /// the specified <paramref name="model"/>.
+        /// </summary>
+        /// <param name="model">The model whose schema elements to cache. This model should be immutable. See <see cref="ExtensionMethods.MarkAsImmutable(IEdmModel)"/>.</param>
+        public CaseInsensitiveSchemaElementsCache(IEdmModel model)
+        {
+            Dictionary<string, List<IEdmSchemaElement>> cache = new Dictionary<string, List<IEdmSchemaElement>>();
+
+            PopulateSchemaElements(model, cache);
+
+            foreach (IEdmModel referencedModel in model.ReferencedModels)
+            {
+                PopulateSchemaElements(referencedModel, cache);
+            }
+
+            this.cache = cache;
+        }
+
+        /// <summary>
+        /// Find all schema elements that match the specified <paramref name="qualifiedName"/>.
+        /// </summary>
+        /// <param name="qualifiedName">The case-insensitive fully qualified name to match.</param>
+        /// <returns>The list of elements that matched the <paramref name="qualifiedName"/>.</returns>
+        public IReadOnlyList<IEdmSchemaElement> FindElements(string qualifiedName)
+        {
+            if (cache.TryGetValue(qualifiedName.ToUpperInvariant(), out List<IEdmSchemaElement> results))

```

Comment:
Avoid these allocations for each look-up.
Make `cache` use StringComparison.OrdinalIgnoreCase comparison instead.

Reply:
Aah, yeah. Thanks.